File: src/data/userStore.ts
import type { User, UserRole } from '../types';
import { isRegistrationRole } from '../config/roles';
import { apiFetch } from '../lib/api';
import { getCachedUsers, reloadCache } from './dataStore';

export interface RoleAccount {
  role: UserRole;
  name: string;
  email: string;
  password: string;
  userId: string;
  status: User['status'];
}

let demoAccountsCache: RoleAccount[] | null = null;

export function getAllUsers(): User[] {
  return getCachedUsers();
}

export function getUsersByRole(role: UserRole): User[] {
  return getAllUsers().filter((u) => u.role === role);
}

export function getUserById(id: string): User | undefined {
  return getAllUsers().find((u) => u.id === id);
}

export async function loadDemoAccounts(): Promise<void> {
  try {
    const { accounts } = await apiFetch<{ accounts: RoleAccount[] }>('/auth/demo-accounts');
    demoAccountsCache = accounts;
  } catch {
    demoAccountsCache = [];
  }
}

export function getRoleAccounts(): RoleAccount[] {
  return demoAccountsCache ?? [];
}

export async function validateLogin(
  email: string,
  password: string
): Promise<{ user: User | null; error?: string }> {
  try {
    const { user } = await apiFetch<{ user: User }>('/auth/login', {
      method: 'POST',
      body: JSON.stringify({ email, password }),
    });
    return { user };
  } catch (err) {
    const message = err instanceof Error ? err.message : 'Login failed';
    return { user: null, error: message };
  }
}

export async function registerUser(
  name: string,
  email: string,
  password: string,
  role: UserRole
): Promise<{ success: boolean; error?: string; user?: User }> {
  if (!isRegistrationRole(role)) {
    return { success: false, error: 'Super Admin accounts cannot be created via registration.' };
  }
  try {
    const { user } = await apiFetch<{ user: User }>('/auth/register', {
      method: 'POST',
      body: JSON.stringify({ name, email, password, role }),
    });
    await reloadCache();
    await loadDemoAccounts();
    return { success: true, user };
  } catch (err) {
    return { success: false, error: err instanceof Error ? err.message : 'Registration failed' };
  }
}

export async function addUser(
  user: Omit<User, 'id' | 'createdAt'>,
  password: string
): Promise<User> {
  const created = await apiFetch<User>('/users', {
    method: 'POST',
    body: JSON.stringify({ user, password, actor: user.name }),
  });
  await reloadCache();
  await loadDemoAccounts();
  return created;
}

export async function updateUser(id: string, data: Partial<User>): Promise<User | null> {
  try {
    const updated = await apiFetch<User>(`/users/${id}`, {
      method: 'PATCH',
      body: JSON.stringify({ data, actor: 'System' }),
    });
    await reloadCache();
    await loadDemoAccounts();
    return updated;
  } catch {
    return null;
  }
}

export async function deleteUser(id: string): Promise<boolean> {
  try {
    await apiFetch(`/users/${id}`, {
      method: 'DELETE',
      body: JSON.stringify({ actor: 'System' }),
    });
    await reloadCache();
    await loadDemoAccounts();
    return true;
  } catch {
    return false;
  }
}

export function setUserPassword(_email: string, _password: string): void {
  /* passwords managed server-side */
}

export function getPasswordForEmail(_email: string): string | undefined {
  return undefined;
}

File: mobiledevelopment/lib/src/data/userStore.dart
import '../lib/api.dart';
import '../config/roles.dart';
import '../types/index.dart';
import 'dataStore.dart';

class RoleAccount {
  final String role;
  final String name;
  final String email;
  final String password;
  final String userId;
  final String status;
  const RoleAccount({required this.role, required this.name, required this.email, required this.password, required this.userId, required this.status});

  factory RoleAccount.fromJson(Map<String, dynamic> json) => RoleAccount(
        role: json['role'] as String,
        name: json['name'] as String,
        email: json['email'] as String,
        password: json['password'] as String,
        userId: json['userId'] as String,
        status: json['status'] as String,
      );
}

class UserStore {
  List<RoleAccount> _demoAccounts = [];

  List<User> getAllUsers() => dataStore.getCachedUsers();
  List<User> getUsersByRole(String role) => getAllUsers().where((u) => u.role == role).toList();
  User? getUserById(String id) {
    try {
      return getAllUsers().firstWhere((u) => u.id == id);
    } catch (_) {
      return null;
    }
  }

  Future<void> loadDemoAccounts() async {
    try {
      final data = await apiFetch<Map<String, dynamic>>('/auth/demo-accounts');
      _demoAccounts = (data['accounts'] as List).map((e) => RoleAccount.fromJson(e)).toList();
    } catch (_) {
      _demoAccounts = [];
    }
  }

  List<RoleAccount> getRoleAccounts() => _demoAccounts;

  Future<({User? user, String? error})> validateLogin(String email, String password) async {
    try {
      final data = await apiFetch<Map<String, dynamic>>('/auth/login', method: 'POST', body: {'email': email, 'password': password});
      return (user: User.fromJson(data['user']), error: null);
    } catch (e) {
      return (user: null, error: e.toString());
    }
  }

  Future<({bool success, String? error, User? user})> registerUser(String name, String email, String password, String role) async {
    if (!isRegistrationRole(role)) {
      return (success: false, error: 'Super Admin accounts cannot be created via registration.', user: null);
    }
    try {
      final data = await apiFetch<Map<String, dynamic>>('/auth/register', method: 'POST', body: {'name': name, 'email': email, 'password': password, 'role': role});
      final user = User.fromJson(data['user']);
      await dataStore.reload();
      await loadDemoAccounts();
      return (success: true, error: null, user: user);
    } catch (e) {
      return (success: false, error: e.toString(), user: null);
    }
  }

  Future<User> addUser(Map<String, dynamic> user, String password) async {
    final data = await apiFetch<Map<String, dynamic>>('/users', method: 'POST', body: {'user': user, 'password': password, 'actor': user['name']});
    await dataStore.reload();
    await loadDemoAccounts();
    return User.fromJson(data);
  }

  Future<User?> updateUser(String id, Map<String, dynamic> data) async {
    try {
      final result = await apiFetch<Map<String, dynamic>>('/users/$id', method: 'PATCH', body: {'data': data, 'actor': 'System'});
      await dataStore.reload();
      await loadDemoAccounts();
      return User.fromJson(result);
    } catch (_) {
      return null;
    }
  }

  Future<bool> deleteUser(String id) async {
    try {
      await apiFetch('/users/$id', method: 'DELETE', body: {'actor': 'System'});
      await dataStore.reload();
      await loadDemoAccounts();
      return true;
    } catch (_) {
      return false;
    }
  }
}

final userStore = UserStore();
